### PR TITLE
support variable-length string type for index

### PR DESCRIPTION
**Pull request by bright-starry-sky**

### CODE REVIEW

**Comment by bright-starry-sky** [comment]
Added override method visit for "CaseExpression" to compile.

File: src/optimizer/OptimizerUtils.cpp [hunk 1]
```diff
@@ -437,5 +437,49 @@ Value OptimizerUtils::boundValueWithMin(const meta::cpp2::ColumnDef& col, const
...
+        case Value::Type::MAP:
+        case Value::Type::DATASET:
+        case Value::Type::PATH: {
+            DLOG(FATAL) << "Not supported value type " << type
+                        << "for index.";
+            return Value(NullType::BAD_TYPE);
+        }
+    }
+    DLOG(FATAL) << "Unknown value type " << static_cast<int>(type);
+    return Value(NullType::BAD_TYPE);
```
Comment (CPWstatic) [inline src/optimizer/OptimizerUtils.cpp:481]: Value::kNullBadType

File: src/validator/MaintainValidator.cpp [hunk 2]
```diff
@@ -303,61 +302,46 @@ Status DropEdgeValidator::toPlan() {
...
-        if (schema_ == nullptr) {
-            return Status::SemanticError("No schema found for '%s'", tagName_.c_str());
-        }
-
-        status = IndexUtil::validateColumns(fields_);
-        if (!status.ok()) {
-            VLOG(1) << status;
-            break;
-        }
-    } while (false);
```
Comment (CPWstatic) [inline src/validator/MaintainValidator.cpp:327]: Why remove these codes?

Comment (bright-starry-sky) [reply src/validator/MaintainValidator.cpp:327]: > Why remove these codes?

No need to check in graph layer, this piece of code is already implemented in the meta layer.
And the meta cache's interval wait is avoided.

File: src/parser/MaintainSentences.cpp [hunk 3]
```diff
@@ -238,9 +238,14 @@ std::string CreateTagIndexSentence::toString() const {
...
     buf += *tagName_;
     buf += " (";
-    std::string columns;
-    folly::join(", ", this->columns(), columns);
-    buf += columns;
+    std::string fields;
+    std::vector<std::string> fieldDefs;
+    for (const auto& field : this->fields()) {
+        std::string f = field.get_name();
+        if (field.__isset.type_length) f + "(" +field.get_type_length() + ")";
```
Comment (yixinglu) [inline src/parser/MaintainSentences.cpp:245]: `fieldDefs.emplace_back(f+...)` ? 

Do you forget to use the string result?

File: src/parser/MaintainSentences.cpp [hunk 4]
```diff
@@ -254,9 +259,14 @@ std::string CreateEdgeIndexSentence::toString() const {
...
     buf += *edgeName_;
     buf += " (";
-    std::string columns;
-    folly::join(", ", this->columns(), columns);
-    buf += columns;
+    std::string fields;
+    std::vector<std::string> fieldDefs;
+    for (const auto& field : this->fields()) {
+        std::string f = field.get_name();
+        if (field.__isset.type_length) f + "(" +field.get_type_length() + ")";
```
Comment (yixinglu) [inline src/parser/MaintainSentences.cpp:266]: ditto

**Review by yixinglu — COMMENTED** [review]
I have a question: Why do we need to specify the length of string column when creating index ?

File: src/parser/MaintainSentences.cpp (hunk 3 above)
Comment (bright-starry-sky) [reply src/parser/MaintainSentences.cpp:244]: > `fieldDefs.emplace_back(f+...)` ?
> 
> Do you forget to use the string result?

Good point. Thanks for reminding. fixed.

**Review by yixinglu — APPROVED** [review]

File: src/optimizer/OptimizerUtils.cpp (hunk 1 above)
Comment (bright-starry-sky) [reply src/optimizer/OptimizerUtils.cpp]: > Value::kNullBadType

Good suggest. unified.

**Review by jievince — APPROVED** [review]
image

File: src/optimizer/OptimizerUtils.cpp [hunk 5]
```diff
@@ -430,11 +430,55 @@ Value OptimizerUtils::boundValueWithMin(const meta::cpp2::ColumnDef& col, const
...
+            return Value::kNullBadType;
         }
     }
     DLOG(FATAL) << "Unknown value type " << static_cast<int>(type);
-    return Value(NullType::BAD_TYPE);
+    return Value::kNullBadType;
+}
+
+Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Value& v) {
+    auto type = SchemaUtil::propTypeToValueType(col.get_type().get_type());
```
Comment (Shylock-Hg) [inline src/optimizer/OptimizerUtils.cpp:441]: Need check does `type` equal to `v.type()`?

File: src/parser/MaintainSentences.cpp [hunk 6]
```diff
@@ -254,9 +260,15 @@ std::string CreateEdgeIndexSentence::toString() const {
     buf += " ON ";
     buf += *edgeName_;
     buf += " (";
-    std::string columns;
-    folly::join(", ", this->columns(), columns);
-    buf += columns;
+    std::vector<std::string> fieldDefs;
+    for (const auto& field : this->fields()) {
```
Comment (Shylock-Hg) [inline src/parser/MaintainSentences.cpp:264]: I think you could move these code to `IndexFieldList::toString()`

File: src/optimizer/OptimizerUtils.cpp (hunk 5 above)
Comment (bright-starry-sky) [reply src/optimizer/OptimizerUtils.cpp:441]: > Need check does `type` equal to `v.type()`?

Not need check at here, the type checked in another method . such as :
```
(root@nebula) [s]> lookup on tag1 where tag1.col1 == "row1_c1"
+------+
| _vid |
+------+
| "v1" |
+------+
Got 1 rows (time spent 1605/1987 us)

Wed, 04 Nov 2020 18:44:55 CST

(root@nebula) [cbs]> lookup on tag1 where tag1.col1 == 1
[ERROR (-12)]: SemanticError: expression error : tag1.col1

Wed, 04 Nov 2020 18:44:59 CST

(root@nebula) [s]> 
```

File: src/parser/parser.yy [hunk 7]
```diff
@@ -1638,14 +1643,46 @@ drop_edge_sentence
...
+        delete $1;
+    }
+    ;
+
+index_field_list
+    : index_field {
+        $$ = new IndexFieldList();
+        std::unique_ptr<meta::cpp2::IndexFieldDef> field;
+        field.reset($1);
+        $$->addField(std::move(field));
```
Comment (Shylock-Hg) [inline src/parser/parser.yy:1668]: I think you could accept raw pointer as parameter directly.

Comment (bright-starry-sky) [reply src/parser/parser.yy:1668]: > I think you could accept raw pointer as parameter directly.

I think here is ok. we don't need to keep this pointer. unique_ptr is enough.

File: src/parser/MaintainSentences.cpp (hunk 6 above)
Comment (bright-starry-sky) [reply src/parser/MaintainSentences.cpp:264]: > I think you could move these code to `IndexFieldList::toString()`

Yeah, I agree with your suggestion, this PR will solve the urgent compilation problem, which I think I will deal with later. thanks for your careful reading.

**Pull request merged**